src/bin/render-readmes.rs

```diff
@@ -0,0 +1,330 @@
+    {
+        let mut req = handle.transfer();
+        req.write_function(|data| {
+            response.extend(data);
+            Ok(data.len())
+        }).unwrap();
+        if let Err(err) = req.perform() {
+            println!(
+                "[{}-{}] Unable to fetch crate: {}",
+                version.krate,
+                version.num,
+                err
+            );
+            return None;
+        }
+    }
+    if handle.response_code().unwrap() != 200 {
+        let response = String::from_utf8_lossy(&response);
+        println!(
+            "[{}-{}] Failed to get a 200 response: {}",
+            version.krate,
+            version.num,
+            response
+        );
+        return None;
+    }
+    let reader = Cursor::new(response);
+    let reader = GzDecoder::new(reader).expect(&format!(
+        "[{}-{}] Invalid gzip header",
+        version.krate,
+        version.num
+    ));
+    let mut archive = Archive::new(reader);
+    let mut entries = archive.entries().expect(&format!(
+        "[{}-{}] Invalid tar archive entries",
+        version.krate,
+        version.num
+    ));
+    let manifest: Manifest = {
+        let path = format!("{}-{}/Cargo.toml", version.krate, version.num);
+        let contents = find_file_by_path(&mut entries, Path::new(&path), &version).unwrap();
+        toml::from_str(&contents).expect(&format!(
+            "[{}-{}] Syntax error in manifest file",
+            version.krate,
+            version.num
+        ))
+    };
+    if manifest.package.readme.is_none() {
+        return None;
+    }
+    let rendered = {
+        let path = format!(
+            "{}-{}/{}",
+            version.krate,
+            version.num,
+            manifest.package.readme.unwrap()
+        );
+        let contents = find_file_by_path(&mut entries, Path::new(&path), &version).unwrap();
+        markdown_to_html(&contents).expect(&format!(
+            "[{}-{}] Couldn't render README",
+            version.krate,
+            version.num
+        ))
+    };
+    return Some(rendered);
+    #[derive(Deserialize)]
+    struct Package {
+        readme: Option<String>,
+    }
+    #[derive(Deserialize)]
+    struct Manifest {
+        package: Package,
+    }
+}
+
+/// Search an entry by its path in a Tar archive.
+fn find_file_by_path<R: Read>(
+    mut entries: &mut tar::Entries<R>,
+    path: &Path,
+    version: &EncodableVersion,
+) -> Option<String> {
+    let mut file = entries
+        .find(|entry| match *entry {
+            Err(_) => false,
+            Ok(ref file) => {
+                let filepath = match file.path() {
+                    Ok(p) => p,
+                    Err(_) => return false,
+                };
+                return filepath == path;
+            }
+        })
+        .expect(&format!(
+            "[{}-{}] file is not present: {}",
+            version.krate,
+            version.num,
+            path.display()
+        ));
+    match file {
```

Comment:
Yeah, I might have done a little bit too much `expect`/`unwrap` for a script supposed to be long-running though it would seem that `panic`'ing in a thread does not affect the main one.

While I'm at it, do we want to refactor the threading model on this script to use a real threadpool and not create a thread for each crate ? It would certainly improve performance but we couldn't `expect`/`unwrap` as much.